cpp/src/fil/treelite_import.cu

```diff
@@ -616,130 +497,138 @@ void tl2fil_common(forest_params_t* params,
-           "but only %d supported for 8-byte sparse nodes",
-           num_features,
-           MAX_FEATURES);
-
-    // check the number of tree nodes
-    const std::vector<tl::Tree<threshold_t, leaf_t>>& trees = model.trees;
-    for (std::size_t i = 0; i < trees.size(); ++i) {
-      int num_nodes = trees[i].num_nodes;
-      ASSERT(num_nodes <= MAX_TREE_NODES,
-             "tree %zu has %d nodes, "
-             "but only %d supported for 8-byte sparse nodes",
-             i,
-             num_nodes,
-             MAX_TREE_NODES);
+    tl2fil_common(&params, model, &tl_params);
+    node_traits<fil_node_t>::check(model);
+
+    size_t num_trees = model.trees.size();
+
+    roots.reserve(num_trees + 1);
+    roots.push_back(0);
+    for (size_t i = 0; i < num_trees; ++i) {
+      int num_nodes =
+        node_traits<fil_node_t>::IS_DENSE ? tree_num_nodes(params.depth) : model.trees[i].num_nodes;
+      roots.push_back(num_nodes + roots.back());
+    }
+    size_t total_nodes = roots.back();
+    roots.pop_back();
+
+    if (params.leaf_algo == VECTOR_LEAF) {
+      size_t max_leaves = node_traits<fil_node_t>::IS_DENSE
+                            ? num_trees * (tree_num_nodes(params.depth) + 1) / 2
+                            : (total_nodes + num_trees) / 2;
+      vector_leaf.resize(max_leaves * params.num_classes);
     }
-  }
-};
 
-// uses treelite model with additional tl_params to initialize FIL params,
-// trees (stored in *ptrees) and sparse nodes (stored in *pnodes)
-template <typename fil_node_t, typename threshold_t, typename leaf_t>
-void tl2fil_sparse(std::vector<int>* ptrees,
-                   std::vector<fil_node_t>* pnodes,
-                   forest_params_t* params,
-                   const tl::ModelImpl<threshold_t, leaf_t>& model,
-                   const treelite_params_t* tl_params,
-                   cat_sets_owner* cat_sets,
-                   std::vector<float>* vector_leaf)
-{
-  tl2fil_common(params, model, tl_params);
-  tl2fil_sparse_check_t<fil_node_t>::check(model);
+    cat_sets = allocate_cat_sets_owner(model);
+    nodes.resize(total_nodes);
 
-  size_t num_trees = model.trees.size();
+// convert the nodes
+#pragma omp parallel for
+    for (std::size_t tree_idx = 0; tree_idx < num_trees; ++tree_idx) {
+      // Max number of leaves processed so far
+      size_t leaf_counter = (roots[tree_idx] + tree_idx) / 2;
+      tree2fil(roots[tree_idx], model.trees[tree_idx], tree_idx, &leaf_counter);
+    }
 
-  ptrees->reserve(num_trees);
-  ptrees->push_back(0);
-  for (size_t i = 0; i < num_trees - 1; ++i) {
-    ptrees->push_back(model.trees[i].num_nodes + ptrees->back());
+    params.num_nodes = nodes.size();
   }
-  size_t total_nodes = ptrees->back() + model.trees.back().num_nodes;
 
-  if (params->leaf_algo == VECTOR_LEAF) {
-    size_t max_leaves = (total_nodes + num_trees) / 2;
-    vector_leaf->resize(max_leaves * params->num_classes);
-  }
+  int tree2fil(int root,
```

Comment:
As discussed offline, let's leave `tree2fil()` outside of `tl2fil_t` for this pull request.

However, this would definitely be good to do in a separate pull request, where the conversions method are moved into `til2fil_t`, and some of them maybe to `tree2fil_t`.